### ISSUE TITLE
Add a HTTP header to make it possible to detect Mastodon

### ISSUE THREAD

**Issue by sindresorhus** [mastodon/mastodon#23939]
### Pitch

Include a HTTP header in every response that indicates that the URL points to a Mastodon instance.

### Motivation

I need to be able to quickly detect whether a link points to a Mastodon instance. I'm aware I could try fetching `https://[instance-url]/api/v1/instance` and deduce it from its content, but that's a bit fragile and also slow.

If I instead could simply perform a HEAD request and check a header, it would be more reliable and faster.

My specific use-case:
I'm the author of a browser picker app called [Velja](https://sindresorhus.com/velja). I want to add support for Mastodon. Meaning that when you click a Mastodon link outside of the browser, it will open with my app and my app will redirect it to a native app like Ivory. The problem is that I cannot know whether a link is a Mastodon link just by looking at the URL. I can have a list of popular Mastodon instances, but I can never cover them all, so I need a quick way to check whether a certain URL points to a Mastodon instance. It needs to be quick as it has to happen right after the user clicks a link and I don't want to keep them waiting.

I imagine there are other use-cases for this too. A link preview scraper could use this to quickly know whether to use a Mastodon scraper for the URL before starting to parse the content. A website that let users add social links could quickly check if the link is a Mastodon link and show the Mastodon icon.

**Comment by Gargron** [comment]
Mastodon returns the “Server” header with the value “Mastodon” but I think it often gets stripped out or overwritten by nginx.

**Comment by voxpelli** [comment]
There is a rel-alternate link of type `application/activity+json` in the HTML response of a page like https://mastodon.social/@sindresorhus/109959846799782156

If that would added to a `Link:` HTTP Header then one could discover it without parsing the HTML response.

(Eventually it could even be added to a 103 Early Hints response or HEAD requests to make it even more lightweight to figure out)

As I noted in https://mastodon.social/@voxpelli/109959930023440575 it would be preferable if the solution is as independent as possible of Mastodon itself, so that other implementations of the same open protocols will also be supported.

**Comment by voxpelli** [comment]
Yeah, the HTTP header is not there now, but the rel-alternate is in the HTML, as the parsed content eg here: https://pin13.net/mf2/?url=https%3A%2F%2Fmastodon.social%2F%40sindresorhus%2F109959846799782156

**Comment by sindresorhus** [comment]
> As I noted in https://mastodon.social/@voxpelli/109959930023440575 it would be preferable if the solution is as independent as possible of Mastodon itself, so that other implementations of the same open protocols will also be supported.

My use-case is to detect Mastodon specifically though, not any fediverse instance. So this feels like a separate issue.

**Comment by bertrandom** [comment]
I am developing a Chrome extension that could redirect to someone's personal Mastodon instance and came across this same issue. The usability issue of clicking on a Mastodon link and not being able to favorite or follow is quite frustrating, I think this would help.

**Comment by voxpelli** [comment]
I was on my phone earlier, now home I could check more properly.

There is a `Link:` header in the response from `HEAD https://mastodon.social/@sindresorhus/109959846799782156`

```http
Link: <https://mastodon.social/users/sindresorhus/statuses/109959846799782156>; rel="alternate"; type="application/activity+json"
```

Parse that using something like https://github.com/Frizlab/LinkHeaderParser and if the `rel` is `alternate` and `type` is `application/activity+json` then you know that there's an alternative representation that's an ActivityStreams JSON document and that is a clear sign that this is content that Ivory can consume.

(This is about the same way that RSS discovery works (but `type` is then eg. `application/atom+xml`)

**Comment by guest20** [comment]
@sinresorhus maybe you can you grab the list of domains from john mastodon's site: https://JoinMastodon.org 

Not a perfect solution, but it might get you part of the way there, and let you do the check without having to make a request from the server at all.

**Comment by sindresorhus** [comment]
@guest20 I'm already doing that. But that only works if I constantly keep the list up to date. I need a fallback for new and unknown servers, which is what this issue would solve.

**Comment by guest20** [comment]
@sindresorhus good thinking :) 

My instances sends both `Server: Mastodon` (mentioned earlier, frontend proxies might replace it with their own name) there's also `Set-Cookie: _mastodon_cookie...`, which seems less likely to be tampered with since changing it on a running instance logs everyone out.

It looks like it was introduced in https://github.com/mastodon/mastodon/commit/ff229aa978d774d9ab067f0cbe345ccbe0f5658c it was requested ... some time ago https://github.com/mastodon/mastodon/issues/141

**Comment by voxpelli** [comment]
Fetch the `application/activity+json` data and introspect it to see if its the kind that should go one way or another.

The document eg. has a [`"type": "Note"`](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-note) when its a toot.

One could also maybe do something like [`jq '.["@context"][].toot?'`](https://jqplay.org/s/_lGCaIqqXF5) on the response to get whether a `toot: "http://joinmastodon.org/ns#"` namespace exists, but that's not really a good way.

**Comment by sindresorhus** [comment]
> Fetch the application/activity+json data and introspect it to see if its the kind that should go one way or another.

That's too slow. I already said I need a HEAD request.

**Comment by mjankowski** [comment]
@sindresorhus can you add more context around why you need to hardcode towards mastodon specifically instead of some of the other more "Feature detection" type suggestions here?

It seems like that approach is risky (ie, the project changes its name and your client is hardcoded), and like some comments said it's a challenge to have some headers not rewritten by http servers on their way out.

Trying to get the full use case though...

**Comment by guest20** [comment]
@mjankowski Knowing that a url points to a Mastodon instance tells the "open with" app that it's worth adding mastodon clients to the list of browsers you can open the url with. 

You wouldn't want to open a link to ebay with a mastodon app, but it would be very handy to let a user choose between opening a mastodon link with the Mastodon app or FediLabs, for example. 

Google Reader is gone, but things still support the google reader API, and mastodon being renamed feels like a pretty similar category of "what if?"

**Comment by mjankowski** [comment]
Thanks for details.

> Knowing that a url points to a Mastodon instance tells the "open with" app that it's worth adding mastodon clients to the list of browsers you can open the url with.

Right - and for most apps this is more easily solved because there's only a few (usually one) relevant hostname to configure ... whereas we in theory have a sprawling empire of hostnames, so we need other mechanisms.

> You wouldn't want to open a link to ebay with a mastodon app

Honestly I would like to do this. There's an `Offer` type in AS - https://www.w3.org/TR/activitystreams-vocabulary/#dfn-offer - how fun would be to keep clicking "bid" in your client of choice and watch it just zap packets of json around while you try to win that 1972 Fiat you've always wanted?!

> but it would be very handy to let a user choose between opening a mastodon link with the Mastodon app or FediLabs, for example.

It sure would, agreed.

> Google Reader is gone, but things still support the google reader API, and mastodon being renamed feels like a pretty similar category of "what if?"

Fair enough.


Is there good prior art for this in other app types or browser extensions that have similar issue? Like maybe an `X-Whatever: ...` http header? (presumably that would be more consistently passed on than `Server` ?)

**Comment by guest20** [comment]
> 1972 Fiat you've always wanted?!

Unfortunately I am one of those 15 minute city types, perhaps there's a nice bike on there?

>  an X-Whatever: ... http header? 

Possibly a [`Link:`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link) pointing at the root of the API? With `rel=mastodon-client`?

**Comment by mjankowski** [comment]
> Unfortunately I am one of those 15 minute city types, perhaps there's a nice bike on there?

Check this out, here's what you could do. First you get a bike with an olde-fashioned bell on it. So you can warn people that you're coming. But, here's the thing, your bell has been cyber-modified to transmit hyper-sonic wireless fidelity waves as you ride around the town. Your bell is both an antenna and a receiver, and its capable of blasting AS-compliant JSON packets all around town. So lets say you and another bike are in close proximity to each other, and the other bike wants to `Offer` to form a little mini-bike-gang with you, even if only for a block or two, that could be negotiated. I guess we've lost ebay from the equation, but it still works.

> Possibly a [`Link:`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link) pointing at the root of the API? With `rel=mastodon-client`?

Maybe?! Is there prior art somewhere for "here's an API of a certain type I provide and here's a URL to find it" that uses either a `Link` header or other `X-` header?

**Comment by guest20** [comment]
https://en.wikipedia.org/wiki/HATEOAS

Though this is only talking about links in a json response, but it seems like a decent analogy.
